### PR TITLE
[KOGITO-692] Allowing specific images to be set for build configurations

### PR DESCRIPTION
**Pull request by ricardozanini**
See:
https://issues.redhat.com/browse/KOGITO-692

In this PR we modified the `KogitoApp` CR to include `imageVersion`, so users could set their preferred version for the builds.

Also, we now accept the full tag name in the `Spec.Build.ImageRuntime|S2I.imageStreamTag` attribute. This will create a new custom `ImageStream` resource with the given tag. See examples in the `deploy/examples/example-custom-build-images.yaml` file.

Needs more tests in the cluster, which I'll do it tomorrow before flying :flight_departure: 

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

File: cmd/kogito/command/deploy/deploy_service.go [hunk 1]
```diff
@@ -149,10 +150,11 @@ func (i *deployCommand) InitHook() {
...
+	i.command.Flags().StringVar(&i.flags.imageRuntime, "image-runtime", "", "Custom image tag for the s2i build, e.g: quay.io/mynamespace/myimage:latest")
 	i.command.Flags().StringVar(&i.flags.installInfinispan, "install-infinispan", defaultInstallInfinispan, "Infinispan installation mode: \"Always\", \"Never\" or \"Auto\". \"Always\" will install Infinispan in the same namespace no matter what, \"Never\" won't install Infinispan even if the service requires it and \"Auto\" will install only if the service requires persistence.")
 	i.command.Flags().StringVar(&i.flags.installKafka, "install-kafka", defaultInstallKafka, "Kafka installation mode: \"Always\" or \"Never\". \"Always\" will use the Strimzi Operator to install a Kafka cluster. The environment variable 'KAFKA_BOOTSTRAP_SERVERS' will be available for the service during runtime.")
+	i.command.Flags().StringVar(&i.flags.imageVersion, "image-version", "", "Image version for standard Kogito build images. Ignored if a custom image is set for image-s2i or image-runtime.")
 }
```
Comment (radtriste) [inline cmd/kogito/command/deploy/deploy_service.go]: no way to specify the `imageStreamNamespace` ?

Comment (radtriste) [inline cmd/kogito/command/deploy/deploy_service.go]: no way to specify the `imageStreamNamespace` ?

**Comment by ricardozanini** [comment]
> If you leave the namespace, why not do like for the version ?
> and then, `imageRuntime` and `imageS2I` have only one attribute, so we could factorize, and at the end have something like that (all at same level):
> 
> ```
> imageVersion: ###
> imageStreamNamespace: ###
> imageRuntimeTag: ###
> imageS2ITag: ###
> ```

TBH I think we should drop the `namespace` attribute.

Doesn't make sense to have the images across namespaces at this moment and we will refactor the build process to work on Kubernetes in the future. The CI use case that I mentioned in the JIRA is still valid, but in that case the IS will be set on the "build" namespace. 

This namespace thing came into a time where we weren't creating IS. So, this should be refactored / dropped when we introduced this feature.

Will be simpler as well to implement and do understand. wdyt?

@spolti @sutaakar, mind to chime in?

**Comment by sutaakar** [comment]
Sounds good to me.

File: cmd/kogito/command/deploy/deploy_service.go (hunk 1 above)
Comment (ricardozanini) [reply cmd/kogito/command/deploy/deploy_service.go]: No way :p

hahaha let's see where our discussions regarding dropping namespace definitely goes.

File: deploy/examples/example-custom-build-images.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    #  value: "http://nexus3.kogito-tools.svc.cluster.local:8081/repository/maven-public/"
+    gitSource:
+      contextDir: jbpm-quarkus-example
+      uri: 'https://github.com/kiegroup/kogito-examples'
+      reference: 0.7.0-rc4
+    # images that don't have a custom tag, will use this version
+    imageVersion: 0.7.0-rc4
+    imageRuntime:
+      # the image runtime build will use this image
+      imageStreamTag: quay.io/ricardozanini/kogito-quarkus-jvm-ubi8:0.7.0-rc1
```
Comment (spolti) [inline deploy/examples/example-custom-build-images.yaml]: maybe use the kiegroup tag 0.7.7-rc4?

File: pkg/apis/app/v1alpha1/zz_generated.deepcopy.go [hunk 3]
```diff
@@ -1,17 +1,3 @@
-// Copyright 2020 Red Hat, Inc. and/or its affiliates
```
Comment (spolti) [inline pkg/apis/app/v1alpha1/zz_generated.deepcopy.go]: do you know why was it removed?

File: deploy/examples/example-custom-build-images.yaml (hunk 2 above)
Comment (ricardozanini) [reply deploy/examples/example-custom-build-images.yaml]: This is to show that we can have a custom tag, thus the namespace naming (and with an actual tag that exists).

File: pkg/apis/app/v1alpha1/zz_generated.deepcopy.go (hunk 3 above)
Comment (ricardozanini) [reply pkg/apis/app/v1alpha1/zz_generated.deepcopy.go]: It's auto-generated. Some times it removes, next commit will add it again.

File: deploy/examples/example-custom-build-images.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    #- name: MAVEN_MIRROR_URL
+    #  value: "http://nexus3.kogito-tools.svc.cluster.local:8081/repository/maven-public/"
+    gitSource:
+      contextDir: jbpm-quarkus-example
+      uri: 'https://github.com/kiegroup/kogito-examples'
+      reference: 0.7.0-rc4
+    # images that don't have a custom tag, will use this version
+    imageVersion: 0.7.0-rc4
+    # the image runtime build will use this image
+    imageRuntimeTag:  quay.io/ricardozanini/kogito-quarkus-jvm-ubi8:0.7.0-rc1
```
Comment (spolti) [inline deploy/examples/example-custom-build-images.yaml]: maybe use kiegroup and tag 0.7.0.rc4?

File: deploy/examples/travelapp-kogito-travel-agency.yaml [hunk 5]
```diff
@@ -17,10 +17,7 @@ spec:
       contextDir: 05-kogito-travel-agency
       reference: override-bootstrap
       uri: https://github.com/ricardozanini/kogito-travel-agency-tutorial
```
Comment (spolti) [inline deploy/examples/travelapp-kogito-travel-agency.yaml]: isn't this example under kiegroup org?

File: pkg/controller/kogitoapp/kogitoapp_controller.go [hunk 6]
```diff
@@ -53,6 +54,11 @@ import (
 
 var log = logger.GetLogger("controller_kogitoapp")
 
+const (
+	// imageStreamNamespace default namespace for the ImageStreams
+	imageStreamNamespace = "openshift"
```
Comment (spolti) [inline pkg/controller/kogitoapp/kogitoapp_controller.go]: Isn't this same env declared somewhere? I do remember to see it but nor sure where exactly

Comment (ricardozanini) [reply pkg/controller/kogitoapp/kogitoapp_controller.go]: This will be removed.

File: deploy/examples/travelapp-kogito-travel-agency.yaml (hunk 5 above)
Comment (ricardozanini) [reply deploy/examples/travelapp-kogito-travel-agency.yaml]: Yes, I'll change it.

File: deploy/examples/example-custom-build-images.yaml (hunk 4 above)
Comment (ricardozanini) [reply deploy/examples/example-custom-build-images.yaml]: The tag is already used by the `version`. The point here is to have a different namespace, so others can understand that they might have their own image in their own namespace.

File: deploy/examples/travelapp-kogito-travel-agency.yaml (hunk 5 above)
Comment (ricardozanini) [reply deploy/examples/travelapp-kogito-travel-agency.yaml]: Actually, only on Maciej's repo. Since it's the source, I'll add it.

**Pull request merged**